### PR TITLE
[FLINK-6834] [table] Support scalar functions on Over Window

### PR DESCRIPTION
**Pull request by wuchong**
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [x] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [x] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [x] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

### CODE REVIEW

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/ProjectionTranslator.scala [hunk 1]
```diff
@@ -226,30 +226,69 @@ object ProjectionTranslator {
...
+            overWindow.get.partitionBy,
+            overWindow.get.orderBy,
+            overWindow.get.preceding,
+            overWindow.get.following)
+        } else {
+          u
+        }
+
+      case u: UnaryExpression =>
+        val c = replaceOverCall(u.child, overWindows, tableEnv)
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/ProjectionTranslator.scala]: Can we add a projection for cover this condition? Such as add `Abs((countFun('b) over 'w))`.
What do you think?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/ProjectionTranslator.scala [hunk 2]
```diff
@@ -226,30 +226,69 @@ object ProjectionTranslator {
...
+      // Scala functions
+      case sfc @ ScalarFunctionCall(clazz, args: Seq[Expression]) =>
+        val newArgs: Seq[Expression] =
+          args.map(
+            (exp: Expression) =>
+              replaceOverCall(exp, overWindows, tableEnv))
+        sfc.makeCopy(Array(clazz, newArgs))
+
+      // Array constructor
+      case c @ ArrayConstructor(args) =>
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/ProjectionTranslator.scala]: Can you explain me that when we using  `ArrayConstructor` ?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/ProjectionTranslator.scala (hunk 1 above)
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/ProjectionTranslator.scala]: Can you explain a bit more what you mean @sunjincheng121 ?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/ProjectionTranslator.scala (hunk 2 above)
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/ProjectionTranslator.scala]: An array constructor would be something like `.select('a, array('b, 'c), 'd)` where `'b` and `'c` would be elements of an array with two fields. Of course, `'b` and `'c` could also be an OVER aggregate.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/ProjectionTranslator.scala (hunk 1 above)
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/ProjectionTranslator.scala]: Ah, I meant the current test case can not execute to this branch， I suggest add `Abs((countFun('b) over 'w))` into test case SELECT clause. i.e:
 .select(
        'a, 'b, 'c,
        'b.sum over 'w,
        "SUM:".toExpr + ('b.sum over 'w),
         countFun('b) over 'w,
         (countFun('b) over 'w) + 1, 
         **Abs((countFun('b) over 'w)), -> // cover UnaryExpression case**
         plusOne(countFun('b) over 'w),
         **array(countFun('c) over 'w, countFun('b) over 'w), -> // cover ArrayConstructor case**
        'b.avg over 'w,
        'b.max over 'w,
        'b.min over 'w,
        weightAvgFun('b, 'a) over 'w)

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/ProjectionTranslator.scala (hunk 2 above)
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/ProjectionTranslator.scala]: Thanks for the explanation  @fhueske :)

**Comment by wuchong** [comment]
@sunjincheng121  thank you for your advice. I add the `array` and `abs` test.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/ProjectionTranslator.scala (hunk 1 above)
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/ProjectionTranslator.scala]: Thanks @sunjincheng121. That makes sense. +1 to that proposal

**Comment by sunjincheng121** [comment]
Thanks for updated. :)
+1 to merged.

**Comment by zentol** [comment]
merging.

**Pull request closed without merging**